File: src/test/java/dev/jiaming/ai_interview/resume/ResumeStorageServiceTests.java
package dev.jiaming.ai_interview.resume;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.nio.charset.StandardCharsets;
import java.util.Map;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.ObjectProvider;

import dev.jiaming.ai_interview.storage.ObjectStorageService;
import dev.jiaming.ai_interview.storage.StoredObject;

class ResumeStorageServiceTests {

	@Test
	void storesNewResumeWithPendingTagAndMarksItReadyAfterExtraction() {
		ObjectStorageService objectStorage = mock(ObjectStorageService.class);
		@SuppressWarnings("unchecked")
		ObjectProvider<ObjectStorageService> provider = mock(ObjectProvider.class);
		when(provider.getIfAvailable()).thenReturn(objectStorage);
		when(objectStorage.put(
			any(), any(), eq("text/plain"),
			eq(Map.of("original-filename", "resume.txt")),
			eq(Map.of("processing-status", "pending"))
		)).thenAnswer(invocation -> {
			String key = invocation.getArgument(0);
			return new StoredObject("bucket", key, 6);
		});
		ResumeStorageService service = new ResumeStorageService(provider);
		byte[] bytes = "resume".getBytes(StandardCharsets.UTF_8);
		ResumeFileContent content = new ResumeFileContent(
			"resume.txt", "text/plain", bytes.length, bytes, "text/plain", "txt"
		);

		String key = service.store(content);
		service.markReady(key);

		assertThat(key).startsWith("resumes/").endsWith("/resume.txt");
		verify(objectStorage).tag(key, Map.of("processing-status", "ready"));
	}
}

File: src/test/kotlin/dev/jiaming/ai_interview/resume/ResumeStorageServiceTests.kt
package dev.jiaming.ai_interview.resume

import dev.jiaming.ai_interview.storage.ObjectStorageService
import dev.jiaming.ai_interview.storage.StoredObject
import java.nio.charset.StandardCharsets
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito
import org.springframework.beans.factory.ObjectProvider

class ResumeStorageServiceTests {
	@Test
	@Suppress("UNCHECKED_CAST")
	fun storesNewResumeWithPendingTagAndMarksItReadyAfterExtraction() {
		val objectStorage = Mockito.mock(ObjectStorageService::class.java)
		val provider = Mockito.mock(ObjectProvider::class.java) as ObjectProvider<ObjectStorageService>
		Mockito.`when`(provider.ifAvailable).thenReturn(objectStorage)
		Mockito.`when`(
			objectStorage.put(
				any(), any(), eq("text/plain"),
				eq(mapOf("original-filename" to "resume.txt")),
				eq(mapOf("processing-status" to "pending"))
			)
		).thenAnswer { invocation ->
			StoredObject("bucket", invocation.getArgument<String>(0), 6)
		}
		val service = ResumeStorageService(provider)
		val bytes = "resume".toByteArray(StandardCharsets.UTF_8)
		val content = ResumeFileContent("resume.txt", "text/plain", bytes.size.toLong(), bytes, "text/plain", "txt")

		val key = service.store(content)
		service.markReady(key)

		assertThat(key).startsWith("resumes/").endsWith("/resume.txt")
		Mockito.verify(objectStorage).tag(key, mapOf("processing-status" to "ready"))
	}
}
